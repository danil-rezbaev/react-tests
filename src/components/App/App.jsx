import './App.scss';
import { useCallback, useState } from "react";
import { Button, Col, Container, Form, Row } from "react-bootstrap";

function App() {
  const defaultValue = {
    operator: '+'
  }

  const [formData, setFormData] = useState(defaultValue)
  const [result, setResult] = useState(undefined)

  const onChangeField = useCallback((event) => {
    const { name, value } = event.target

    setFormData((data) => ({
        ...data,
        [name]: value
      }))
  }, [])

  const calculate = useCallback(() => {
    const {operator, firstArg, secondArg} = formData

    console.log('formData', formData)

    if (!firstArg || !secondArg || !operator) {
      return 0
    }

    if (typeof +firstArg !== "number" || typeof +secondArg !== "number") {
      return 0
    }

    if(operator === '+') {
      return +firstArg + +secondArg
    } else if(operator === '-') {
      return +firstArg - +secondArg
    } else if(operator === '*') {
      return +firstArg * +secondArg
    }
  }, [formData])

  const buttonClick = useCallback((event) => {
    const form = event.currentTarget;

    event.preventDefault();
    event.stopPropagation();

    if (form.checkValidity() === true) {
      setResult(calculate())
    }
  }, [calculate])

  return (
    <div className="App">
      <Container className="mt-5">
        <Row>
          <Col>
            <Form.Control
              placeholder="аргумент 1"
              name="firstArg"
              onChange={onChangeField}
              value={formData["firstArg"] ?? ''}
            />
          </Col>

          <Col>
            <Form.Select
              name="operator"
              onChange={onChangeField}
              aria-label="оператор"
              value={formData["operator"] ?? '+'}
              role="listbox"
              required
            >
              <option value="+">+</option>
              <option value="-">-</option>
              <option value="*">*</option>
            </Form.Select>
          </Col>

          <Col>
            <Form.Control
              placeholder="аргумент 2"
              name="secondArg"
              onChange={onChangeField}
              value={formData["secondArg"] ?? ''}
              className="form-control"
            />
          </Col>

          <Col>
            { typeof result !== 'undefined'
              ? <h4 className="result text-start">= {result}</h4>
              : null
            }
          </Col>

          <Col md={12} className="mt-2">
            <Button
              type="submit"
              variant="primary"
              className="text-white w-100"
              role="submitButton"
              onClick={buttonClick}
            >
              Рассчитать!
            </Button>
          </Col>
        </Row>
      </Container>
    </div>
  );
}

export default App;
