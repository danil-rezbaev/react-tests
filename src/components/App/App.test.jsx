import { render, screen } from '@testing-library/react';
import App from './App';
import userEvent from "@testing-library/user-event";

describe("App component", () => {
  test('app component renders', () => {
    render(<App />);
  });

  test('app component render with unnecessary props', () => {
    render(<App data="prop" />);
  });

  test('result default be null', () => {
    render(<App />);
    expect(screen.queryByText(/=/)).toBeNull()
  });

  test('fields are on the page', () => {
    const { container } = render(<App />);

    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const formĞ¡ontrols = container.querySelectorAll('.form-control');
    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const formSelects = container.querySelectorAll('.form-select');

    expect(formĞ¡ontrols.length).toBe(2);
    expect(formSelects.length).toBe(1);
  });

  test('submit button are on the page', () => {
    const { container } = render(<App />);

    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const buttonSubmit = container.querySelector('button[type="submit"]');

    expect(buttonSubmit).toBeInTheDocument()
  });

  test('the value of the .form-control changes when typing', () => {
    const { container } = render(<App />);
    const textToType = '10'

    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const firstArg = container.querySelector(`input[name="firstArg"]`);

    userEvent.type(firstArg, textToType)
    expect(firstArg).toHaveAttribute('value', textToType)
  });

  test('test select has default value', () => {
    render(<App />);
    expect(screen.getByText("+")).toBeInTheDocument();
  });

  test('changing select values works as expected', () => {
    render(<App />);

    userEvent.selectOptions(screen.getByRole('listbox'), ['*'])

    expect(screen.getByRole('option', {name: '+'}).selected).toBe(false)
    expect(screen.getByRole('option', {name: '-'}).selected).toBe(false)
    expect(screen.getByRole('option', {name: '*'}).selected).toBe(true)
  });

  test('testing the addition of numbers', () => {
    const { container } = render(<App />);
    const textToType = '10'
    userEvent.selectOptions(screen.getByRole('listbox'), ['+'])

    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const firstArg = container.querySelector(`input[name="firstArg"]`);
    userEvent.type(firstArg, textToType)

    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const secondArg = container.querySelector(`input[name="secondArg"]`);
    userEvent.type(secondArg, textToType)

    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const buttonSubmit = container.querySelector(`button[type="submit"]`);
    userEvent.click(buttonSubmit)

    expect(screen.getByText(/20/)).toBeInTheDocument();
  });

  test('testing the subtraction of numbers', () => {
    const { container } = render(<App />);
    const textToType = '10'
    userEvent.selectOptions(screen.getByRole('listbox'), ['-'])

    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const firstArg = container.querySelector(`input[name="firstArg"]`);
    userEvent.type(firstArg, textToType)

    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const secondArg = container.querySelector(`input[name="secondArg"]`);
    userEvent.type(secondArg, textToType)

    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const buttonSubmit = container.querySelector(`button[type="submit"]`);
    userEvent.click(buttonSubmit)

    expect(screen.getByText(/0/)).toBeInTheDocument();
  });

  test('testing the multiplication of numbers', () => {
    const { container } = render(<App />);
    const textToType = '10'
    userEvent.selectOptions(screen.getByRole('listbox'), ['*'])

    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const firstArg = container.querySelector(`input[name="firstArg"]`);
    userEvent.type(firstArg, textToType)

    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const secondArg = container.querySelector(`input[name="secondArg"]`);
    userEvent.type(secondArg, textToType)

    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const buttonSubmit = container.querySelector(`button[type="submit"]`);
    userEvent.click(buttonSubmit)

    expect(screen.getByText(/100/)).toBeInTheDocument();
  });

  test('App snapshot', () => {
    const view = render(<App />)
    expect(view).toMatchSnapshot()
  })
})
