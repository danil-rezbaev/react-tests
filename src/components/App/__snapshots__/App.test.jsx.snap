// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component App snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App"
      >
        <div
          class="mt-5 container"
        >
          <div
            class="row"
          >
            <div
              class="col"
            >
              <input
                class="form-control"
                name="firstArg"
                placeholder="аргумент 1"
                value=""
              />
            </div>
            <div
              class="col"
            >
              <select
                aria-label="оператор"
                class="form-select"
                name="operator"
                required=""
                role="listbox"
              >
                <option
                  value="+"
                >
                  +
                </option>
                <option
                  value="-"
                >
                  -
                </option>
                <option
                  value="*"
                >
                  *
                </option>
              </select>
            </div>
            <div
              class="col"
            >
              <input
                class="form-control form-control"
                name="secondArg"
                placeholder="аргумент 2"
                value=""
              />
            </div>
            <div
              class="col"
            />
            <div
              class="mt-2 col-md-12"
            >
              <button
                class="text-white w-100 btn btn-primary"
                role="submitButton"
                type="submit"
              >
                Рассчитать!
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App"
    >
      <div
        class="mt-5 container"
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <input
              class="form-control"
              name="firstArg"
              placeholder="аргумент 1"
              value=""
            />
          </div>
          <div
            class="col"
          >
            <select
              aria-label="оператор"
              class="form-select"
              name="operator"
              required=""
              role="listbox"
            >
              <option
                value="+"
              >
                +
              </option>
              <option
                value="-"
              >
                -
              </option>
              <option
                value="*"
              >
                *
              </option>
            </select>
          </div>
          <div
            class="col"
          >
            <input
              class="form-control form-control"
              name="secondArg"
              placeholder="аргумент 2"
              value=""
            />
          </div>
          <div
            class="col"
          />
          <div
            class="mt-2 col-md-12"
          >
            <button
              class="text-white w-100 btn btn-primary"
              role="submitButton"
              type="submit"
            >
              Рассчитать!
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
